### PR TITLE
wasm2c: Add segue option to make exclusive use of the segment register

### PR DESCRIPTION
**Pull request by shravanrn**
The 3rd of the 3 PRs needed for production Linux Segue support (i.e., use x86 segment registers to refer to the Wasm heap for single-memory Wasm modules). This PR adds an option for applications to tell wasm2c that it has exclusive use of the gs register for this purpose and need not save and restore it on each use --- something that is particularly expensive on pre-Ivybridge CPUs which requires syscalls to set segment registers.

### CODE REVIEW

**Comment by shravanrn** [comment]
@sbc100 Please have a look when you have the chance. This should be the last segue commit prior to us deploying in Firefox.

File: test/wasm2c/minimal.txt [hunk 1]
```diff
@@ -119,10 +119,21 @@ static inline void* wasm_rt_segue_read_base() {
   }
 }
 static inline void wasm_rt_segue_write_base(void* base) {
-  if (wasm_rt_fsgsbase_inst_supported) {
-    __builtin_ia32_wrgsbase64((uintptr_t)base);
+  bool should_set = true;
+#if WASM_RT_SEGUE_FREE_SEGMENT
+  if (wasm_rt_last_segment_val == base) {
+    should_set = false;
```
Comment (sbc100) [inline test/wasm2c/minimal.txt:125]: Can we replace this line with `return;` and the avoid the `should_set` variable completely?

Comment (shravanrn) [reply test/wasm2c/minimal.txt]: Fixed! You're right, that's way better :)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -203,7 +203,7 @@ jobs:
     env:
       USE_NINJA: "1"
       WASM2C_CC: "clang"
-      WASM2C_CFLAGS: "-DWASM_RT_USE_MMAP=1 -DWASM_RT_SKIP_SIGNAL_RECOVERY=1 -DWASM_RT_NONCONFORMING_UNCHECKED_STACK_EXHAUSTION=1 -DWASM2C_TEST_EMBEDDER_SIGNAL_HANDLING -DWASM_RT_ALLOW_SEGUE=1 -mfsgsbase -DWASM_RT_SANITY_CHECKS=1 -Wno-pass-failed"
+      WASM2C_CFLAGS: "-DWASM_RT_USE_MMAP=1 -DWASM_RT_SKIP_SIGNAL_RECOVERY=1 -DWASM_RT_NONCONFORMING_UNCHECKED_STACK_EXHAUSTION=1 -DWASM2C_TEST_EMBEDDER_SIGNAL_HANDLING -DWASM_RT_ALLOW_SEGUE=1 -DWASM_RT_SEGUE_FREE_SEGMENT=1 -mfsgsbase -DWASM_RT_SANITY_CHECKS=1 -Wno-pass-failed"
```
Comment (sbc100) [inline .github/workflows/build.yml:206]: On the name of this new macro, would it make sense to call it something like `WASM_RT_SEGUE_PERSISTENT` ?

File: wasm2c/wasm-rt-impl.c [hunk 3]
```diff
@@ -62,6 +62,10 @@ bool wasm_rt_fsgsbase_inst_supported = false;
 #include <unistd.h>       // For syscall
 #endif
 
+#if WASM_RT_SEGUE_FREE_SEGMENT
+WASM_RT_THREAD_LOCAL void* wasm_rt_last_segment_val = 0;
```
Comment (sbc100) [inline wasm2c/wasm-rt-impl.c:66]: NULL?

**Review by sbc100 — APPROVED** [review]
lgtm % minor comments.

File: .github/workflows/build.yml (hunk 2 above)
Comment (sbc100) [reply .github/workflows/build.yml:206]: Having read the change again, I'm not sure my name is any better.  lgtm either way.

Comment (shravanrn) [reply .github/workflows/build.yml:206]: Ah ok, if no preference, I'll leave it as is.

Comment (SoniEx2) [reply .github/workflows/build.yml:206]: (hmm, what about `SEGUE_SEGMENT_EXCLUSIVE`?)

**Pull request merged**